### PR TITLE
Issue1502 replaceable thermal zone

### LINKED ISSUE #1502: Implement a thermal zone model as a replaceable class

### PR DESCRIPTION
**Pull request by milicag**
Closes #1502

### CODE REVIEW

**Comment by milicag** [comment]
@mwetter This is ready for review.

File: Buildings/Experimental/EnergyPlus/Examples/VAVReheatRefBldgSmallOffice/Guideline36_EnergyPlusZone.mo [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+<ul>
+<li>
+November 25, 2019, by Milica Grahovac:<br/>
+Impementation of <a href=\"modelica://Buildings.Examples.VAVReheat.Guideline36\">
+Buildings.Examples.VAVReheat.Guideline36</a> model with an EnergyPlus thermal zone instance.
+</li>
+</ul>
+</html>"),
+    __Dymola_Commands(file=
+          "modelica://Buildings/Resources/Scripts/Dymola/Examples/VAVReheat/Guideline36.mos"
```
Comment (mwetter) [inline Buildings/Experimental/EnergyPlus/Examples/VAVReheatRefBldgSmallOffice/Guideline36_EnergyPlusZone.mo]: That is the wrong script.

**Comment by milicag** [comment]
@mwetter I have not updated the reference results for the two VAVReheat example models so you can take a look. `PartialFloor` lives under `Buildings.ThermalZones.EnergyPlus.Examples.VAVReheatRefBldgSmallOffice.BaseClasses` and potentially could move somewhere more reasonable, as the model is being extended by `Buildings.Examples.VAVReheat.ThermalZones.Floor` as well. As discussed on Tuesday, I'm handing this to you unfinished, but let me know if you'd like any edits.

**Comment by mwetter** [comment]
@milicag : Can you please make sure all scripts are committed, for example, `Resources/Scripts/Dymola/Experimental/EnergyPlus/Examples/VAVReheatRefBldgSmallOffice/Guideline36.mos` is missing. I will need these to review and work further on it.

**Comment by milicag** [comment]
@mwetter Those models now live under `ThermalZones`, as you've moved the `EnergyPlus` package over, so the script I believe you are looking for is [already included](https://github.com/lbl-srg/modelica-buildings/pull/1663/files#diff-30b0d6a33fd34011cc81a3fe9a1cc96e). I did see a small error so let me commit a fix.

**Comment by mwetter** [comment]
If you run check model, the check will fail. Please update the path to the
script so that the model checks are successful.

**Comment by milicag** [comment]
@mwetter How do you run check model?

**Comment by mwetter** [comment]
@milicag there is a button in the Dymola Graphic or Text view.

**Comment by milicag** [comment]
@mwetter I run the checks, let me know if you need anything else.

File: Buildings/Resources/Scripts/Dymola/ThermalZones/EnergyPlus/Examples/VAVReheatRefBldgSmallOffice/FloorOpenLoop.mos [hunk 2]
```diff
@@ -1,5 +1,5 @@
 old_hidden_avoid_double_computation=Hidden.AvoidDoubleComputation;
 Hidden.AvoidDoubleComputation=true;
-simulateModel("Buildings.ThermalZones.EnergyPlus.Examples.VAVReheatRefBldgSmallOffice.FloorOpenLoopFrom1129", stopTime=172800, method="CVode", tolerance=1e-06, resultFile="FloorOpenLoop");
+simulateModel("Buildings.ThermalZones.EnergyPlus.Examples.VAVReheatRefBldgSmallOffice.FloorOpenLoop", stopTime=172800, method="Radau", tolerance=1e-06, resultFile="FloorOpenLoop");
```
Comment (milicag) [inline Buildings/Resources/Scripts/Dymola/ThermalZones/EnergyPlus/Examples/VAVReheatRefBldgSmallOffice/FloorOpenLoop.mos]: Sorry about that.

File: Buildings/ThermalZones/EnergyPlus/Examples/VAVReheatRefBldgSmallOffice/BaseClasses/Floor.mo [hunk 3]
```diff
@@ -66,6 +66,16 @@ model Floor "Model of a floor of the building"
   Modelica.Blocks.Sources.Constant qLatGai_flow(k=0) "Latent heat gain"
     annotation (Placement(transformation(extent={{214,380},{234,400}})));
 
+protected
+  inner Building building(
```
Comment (milicag) [inline Buildings/ThermalZones/EnergyPlus/Examples/VAVReheatRefBldgSmallOffice/BaseClasses/Floor.mo]: Elegant.

**Comment by milicag** [comment]
@mwetter Please let me know if you need anything from me on this at this point.

**Comment by mwetter** [comment]
@milicag This is fine for now but needs more testing after I get an updated binary for the Spawn envelope.

**Comment by milicag** [comment]
@mwetter Please see the validation test run error I am getting. I will look further into it tomorrow and please let me know if you have any pointers.

```
Translation of Buildings.ThermalZones.EnergyPlus.Examples.VAVReheatRefBldgSmallOffice.FloorOpenLoop:

The DAE has 2951 scalar unknowns and 2951 scalar equations.

Statistics

Selected continuous time states

Compilation of Buildings.ThermalZones.EnergyPlus.Examples.VAVReheatRefBldgSmallOffice.FloorOpenLoop failed.

Compiler message:
In file included from /opt/dymola-2021-x86_64/source/dsblock1.c:68:0,
from dsmodel.c:1452:
/opt/dymola-2021-x86_64/source/ModelicaStrings.c: In function ‘ModelicaStrings_scanInteger’:
/opt/dymola-2021-x86_64/source/ModelicaStrings.c:315:26: warning: implicit declaration of function ‘strtol_l’; did you mean ‘strtoll’? [-Wimplicit-function-declaration]
x = (int)strtol_l(buf, &endptr, 10, loc);
^~~~~~~~
strtoll
/opt/dymola-2021-x86_64/source/ModelicaStrings.c: In function ‘ModelicaStrings_scanReal’:
/opt/dymola-2021-x86_64/source/ModelicaStrings.c:427:13: warning: implicit declaration of function ‘strtod_l’; did you mean ‘strtoull’? [-Wimplicit-function-declaration]
x = strtod_l(buf, &endptr, loc);
^~~~~~~~
strtoull
In file included from /opt/dymola-2021-x86_64/source/matrixop.c:3855:0:
/opt/dymola-2021-x86_64/source/localeless.cpp: In function ‘fscanfC_Bracket_255s’:
/opt/dymola-2021-x86_64/source/localeless.cpp:250:2: warning: ignoring return value of ‘fscanf’, declared with attribute warn_unused_result [-Wunused-result]
fscanf(f, "%[^\\[]255s", s);
^~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/bin/ld:/home/mg/repos/modelica-buildings/Buildings/Resources/Library/linux64/libModelicaBuildingsEnergyPlus.so: file format not recognized; treating as linker script
/usr/bin/ld:/home/mg/repos/modelica-buildings/Buildings/Resources/Library/linux64/libModelicaBuildingsEnergyPlus.so:1: syntax error
collect2: error: ld returned 1 exit status

ERRORS have been issued.
```

**Comment by mwetter** [comment]
@milicag The message
```
/usr/bin/ld:/home/mg/repos/modelica-buildings/Buildings/Resources/Library/linux64/libModelicaBuildingsEnergyPlus.so: file format not recognized; treating as linker script
```
looks like your system did not check out the binary, which can happen if git-lfs is not installed.

**Comment by milicag** [comment]
@mwetter I think this is good to go, then @AntoineGautier 's branch can be evaluated, unless I'm missing something. Please let me know. 

Is `git lfs` info for those who clone the repo already somewhere in the install documentation? I looked around on the wiki a bit and was not able to find it - I could add it to https://github.com/lbl-srg/modelica-buildings/wiki/Git-Information.

**Comment by milicag** [comment]
@mwetter Please note that the test fails for `Guideline36` model due to variation in the reference results. I was reluctant to update before I better understand the intent of @AntoineGautier [branch](https://github.com/lbl-srg/modelica-buildings/pull/2021) in relation to this.

**Comment by mwetter** [comment]
@milicag I updated https://github.com/lbl-srg/modelica-buildings/wiki/Git-Information

@AntoineGautier : Can you please https://github.com/lbl-srg/modelica-buildings/pull/1663#issuecomment-701154538 when you are back next week so we don't overwrite your changes.

**Comment by milicag** [comment]
Thanks @mwetter and @AntoineGautier.

**Comment by AntoineGautier** [comment]
The fast flow variations come from #2017 and not from the organic changes tracked in #2021. 
See the plot below where the heating demand signal `uHea` is low but the minimum air flow set point in heating mode `vActHeaMin_flow` is forced to the nominal value when `uOpeMod = 4` (Warm-up). 

RoomVAV

@JayHuLBL Note that this is because the default value of the parameter `VDisHeaSetMax_flow` in `Buildings.Controls.OBC.ASHRAE.G36_PR1.TerminalUnits.Controller` is set to `V_flow_nominal`. And it is not overwritten in the example models.
I wonder if we should not set it to a more realistic value, for instance `0.3 *  V_flow_nominal`, at least for the example models (if G36 states that by default it should be equal to the cooling design flow rate). 
The value  `0.3 *  V_flow_nominal` is the one used as the maximum flow rate in heating mode in `ASHRAE2006` so that will make the two models more consistent and their comparison more fair.

**Comment by milicag** [comment]
Based on team discussion resulting in creation of [this issue](https://github.com/lbl-srg/modelica-buildings/issues/2172) the flow spikes during warm up operation result in, with the current G36 secondary sequences, the expected behavior and the same effect @AntoineGautier has observed in the example models that use our library's native floor models. The above linked issue will address the default values in a separate PR to reduce the magnitude of the spikes with a goal to better match common design approach in practice and also better separate the characteristics of the controls implementation from the specifications of the equipment controlled.

With this I will update the reference results for the EnergyPlus G36 example.

**Comment by milicag** [comment]
I intend to replace this PR with a new one [based on this branch](https://github.com/lbl-srg/modelica-buildings/tree/issue1502_new_add_of_replaceable_thermal_zone_models).

**Comment by milicag** [comment]
Closing this PR due to an unusually large file diff that the team (Milica, Antoin) was not able to observe locally.

**Pull request closed without merging**